firebase-perf/src/test/java/com/google/firebase/perf/util/FirstDrawDoneListenerTest.java

```diff
@@ -0,0 +1,123 @@
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+/** Unit tests for {@link FirstDrawDoneListener}. */
+@RunWith(RobolectricTestRunner.class)
+public class FirstDrawDoneListenerTest {
+  @Mock private Handler handler;
+  @Captor private ArgumentCaptor<Runnable> callbackCaptor;
+
+  private View testView;
+
+  @Before
+  public void setUp() {
+    initMocks(this);
+    testView = new View(ApplicationProvider.getApplicationContext());
+  }
+
+  @Test
+  @Config(sdk = 25)
+  public void registerForNextDraw_delaysAddingListenerForAPIsBelow26()
+      throws NoSuchFieldException, IllegalAccessException {
+    // Add a listener first so ViewTreeObserver.mOnDrawListeners is initialized and non-null
+    testView.getViewTreeObserver().addOnDrawListener(() -> {});
+    Field mOnDrawListeners =
+        android.view.ViewTreeObserver.class.getDeclaredField("mOnDrawListeners");
+    mOnDrawListeners.setAccessible(true);
+    ArrayList<ViewTreeObserver.OnDrawListener> listOfListeners =
+        (ArrayList<ViewTreeObserver.OnDrawListener>)
+            mOnDrawListeners.get(testView.getViewTreeObserver());
+    assertThat(listOfListeners).isNotNull();
+    assertThat(listOfListeners.size()).isEqualTo(1);
+
+    // OnDrawListener is not registered, it is delayed for later
+    FirstDrawDoneListener.registerForNextDraw(testView, () -> {});
+    assertThat(listOfListeners.size()).isEqualTo(1);
+  }
+
+  @Test
+  @Config(sdk = 26)
+  public void registerForNextDraw_directlyAddsListenerForApi26AndAbove()
+      throws NoSuchFieldException, IllegalAccessException {
+    // Add a listener first so ViewTreeObserver.mOnDrawListeners is initialized and non-null
+    testView.getViewTreeObserver().addOnDrawListener(() -> {});
+    Field mOnDrawListeners =
+        android.view.ViewTreeObserver.class.getDeclaredField("mOnDrawListeners");
+    mOnDrawListeners.setAccessible(true);
+    ArrayList<ViewTreeObserver.OnDrawListener> listOfListeners =
+        (ArrayList<ViewTreeObserver.OnDrawListener>)
+            mOnDrawListeners.get(testView.getViewTreeObserver());
+    assertThat(listOfListeners).isNotNull();
+    assertThat(listOfListeners.size()).isEqualTo(1);
+
+    // Immediately register an OnDrawListener to ViewTreeObserver
+    FirstDrawDoneListener.registerForNextDraw(testView, () -> {});
+    assertThat(listOfListeners.size()).isEqualTo(2);
+  }
+
+  @Test
+  public void onDraw_postsCallbackToFrontOfQueue() {
+    Runnable callback = () -> {};
+    testView
+        .getViewTreeObserver()
+        .addOnDrawListener(new FirstDrawDoneListener(testView, callback, handler));
+    testView.getViewTreeObserver().dispatchOnDraw();
+    verify(handler).postAtFrontOfQueue(callbackCaptor.capture());
+    assertThat(callbackCaptor.getValue()).isSameInstanceAs(callback);
+  }
+
+  @Test
+  @Config(sdk = 26)
```

Comment:
Can we have a test for un-registering itself in Config(sdk=25) for both OnAttachStateChangeListener and OnDrawListener?

Reply:
I only added `@Config(sdk = 26)` for this test because it uses reflection, so I don't know if in the future Android will change field names or refactor `ViewTreeObserver`. 

The < API 26 bug is that adding `OnDrawListeners` to `ViewTreeObserver` doesn't work, it has nothing to do with removing or the listener itself, so I don't think we need to test unregistering for 2 different versions.

Reply:
For the < API 26 bug, we should be making sure that we have a test to make sure that if we do register a `View.OnAttachStateChangeListener`, that it is unregistered once `onViewAttachedToWindow` is called. Otherwise, we could change the code to leave it registered and the tests would still pass.

Reply:
Ah, I see what you mean. That is difficult for the same reason as the other comment: `View` class doesn't have public, easy to use `dispatchXYZ()` methods like `ViewTreeObserver` does.

Reply:
Could you use a FakeView or mock the View object? This way you have access to the listeners and can trigger the interfaces directly.

Reply:
Added check at the last line of `registerForNextDraw_delaysAddingListenerForAPIsBelow26()`.